robot_smach_states/src/robot_smach_states/navigation/guidance.py

```diff
@@ -30,30 +34,183 @@ def _detect_operator_behind_robot(robot, distance=1.0, radius=0.5):
+            (person.pose.frame.p - expected_person_pos).Norm(),
+            radius
+        ))
         if (person.pose.frame.p - expected_person_pos).Norm() < radius:
             return True
     return False
 
 
+class TourGuide(object):
+    def __init__(self, robot, x_threshold=0.75, y_threshold=1.5):
+
+        self._robot = robot
+
+        # Properties
+        self._x_threshold = x_threshold
+        self._y_threshold = y_threshold
+
+        # State
+        self._furniture_entities = []
+        self._room_entities = []
+        self._furniture_entities_room = {}  # map room entities to furniture entities
+        self._passed_room_ids = []  # Will contain the ids of the rooms that are passed
+        self._passed_furniture_ids = []  # Will contain the ids of the furniture that is passed
+
+        self.initialize()
+
+    def describe_near_objects(self):
+        # type: () -> string
+        """
+        Describes near objects based on the robots current position
+        :return: A description of the robots surroundings
+        :rtype: string
+        """
+        position = self._robot.base.get_location()
+
+        try:
+            room = self.get_room(position.frame.p)
+        except RuntimeError:
+            rospy.logwarn("position ({}) not in any room".format(position.frame.p))
+            return ""
+
+        if room.id not in self._passed_room_ids:
+            self._passed_room_ids.append(room.id)
+            rospy.logdebug("describe entering room: {}.\t passed rooms is now {}".format(room.id, self._passed_room_ids))
+            return "We now enter the {}".format(room.id)
+
+        # not entering a new room, checking furniture
+        furniture_objects = self._furniture_entities_room[room]  # Furniture objects in our current room
+        for entity in furniture_objects:  # type: Entity
+            # Check if in passed ids
+            if entity.id in self._passed_furniture_ids:
+                continue
+
+            # Compute the pose of the entity w.r.t. the 'path'
+            entity_relative_pose = position.frame.Inverse() * entity.pose.frame
+
+            # Check the distance
+            if abs(entity_relative_pose.p.x()) < self._x_threshold and abs(entity_relative_pose.p.y()) < self._y_threshold:
+                self._passed_furniture_ids.append(entity.id)
+                side = "left" if entity_relative_pose.p.y() >= 0.0 else "right"
+                rospy.logdebug(
+                    "describe passing entity: {}.\t passed entities is now {}".format(room.id, self._passed_room_ids))
+                return "On our {} you can see the {}".format(side, entity.id)
+
+        # no furniture passed, nothing of interest
+        return ""
+
+    def initialize(self):
+        entities = self._robot.ed.get_entities()
+        self._furniture_entities = [entity for entity in entities if entity.is_a("furniture")]
+        self._room_entities = [room for room in entities if room.type == "room"]
```

Comment:
How does this work @jlunenburg? Where is room specified?

Reply:
In ed_object models. A room is also an entity with the type "room"